dids/diddht/internal/dns/did.go

```diff
@@ -0,0 +1,278 @@
+		props = append(props, "c="+vm.Controller)
+	}
+
+	dhtEncodedVM := strings.Join(props, ";")
+
+	return dhtEncodedVM, nil
+
+}
+
+// MarshalService packs a service into a TXT DNS resource record and adds to the DNS message Answers
+func MarshalService(dhtDNSkey string, s *didcore.Service, msg *dnsmessage.Message) error {
+	rawData := fmt.Sprintf("id=%s;t=%s;se=%s", s.ID, s.Type, s.ServiceEndpoint)
+
+	resource, err := newResource(fmt.Sprintf("_%s._did.", dhtDNSkey), rawData)
+	if err != nil {
+		return err
+	}
+	msg.Answers = append(msg.Answers, resource)
+	return nil
+}
+
+// UnmarshalVerificationMethod unpacks the TXT DNS resource encoded verification method
+func UnmarshalVerificationMethod(data string, vm *didcore.VerificationMethod) error {
+	propertyMap, err := parseTXTRecordData(data)
+	if err != nil {
+		return err
+	}
+
+	vm.Type = "JsonWebKey"
+
+	var key string
+	var algorithmID string
+	for property, v := range propertyMap {
+		switch property {
+		// According to https://did-dht.com/#verification-methods, this should not be a list
+		case "id":
+			vm.ID = strings.Join(v, "")
+		case "t": // Index of the key type https://did-dht.com/registry/index.html#key-type-index
+			algorithmID, _ = dhtIndexToAlg[strings.Join(v, "")]
+		case "k": // unpadded base64URL representation of the public key
+			key = strings.Join(v, "")
+		case "c": // the controller is optional
+			vm.Controller = strings.Join(v, "")
+		default:
+			continue
+		}
+	}
+
+	if len(key) == 0 || len(algorithmID) == 0 {
+		return errors.New("unable to parse public key")
+	}
+
+	// RawURLEncoding is the same as URLEncoding but omits padding.
+	// Decoding and reencoding to make sure there is no padding
+	keyBytes, err := base64.RawURLEncoding.DecodeString(key)
+	if err != nil {
+		return err
+	}
+
+	if len(keyBytes) == 0 {
+		return errors.New("malformed public key")
+	}
+
+	j, err := dsa.BytesToPublicKey(algorithmID, keyBytes)
+	if err != nil {
+		return err
+	}
+	vm.PublicKeyJwk = &j
+
+	// validate all the parts exist
+	if len(vm.ID) == 0 || vm.PublicKeyJwk == nil {
+		return errors.New("malformed verification method representation")
+	}
+
+	return nil
+}
+
+// UnmarshalService unpacks the TXT DNS resource encoded service
+func UnmarshalService(data string, s *didcore.Service) error {
+	propertyMap, err := parseTXTRecordData(data)
+	if err != nil {
+		return err
+	}
+	for property, v := range propertyMap {
+		switch property {
+		case "id":
+			s.ID = strings.Join(v, "")
+		case "t":
+			s.Type = strings.Join(v, "")
+		case "se":
+			validEndpoints := []string{}
+			for _, uri := range v {
+				if _, err := url.ParseRequestURI(uri); err != nil {
+					return errors.New("invalid service endpoint")
+				}
+				validEndpoints = append(validEndpoints, uri)
+			}
+			s.ServiceEndpoint = strings.Join(validEndpoints, ",")
+		default:
+			continue
+		}
+	}
+
+	return nil
+}
+
+func pluckSort(hayStack []didcore.VerificationMethod) []string {
+	ids := []string{}
```

Comment:
nit: https://github.com/uber-go/guide/blob/master/style.md#use-var-for-zero-value-structs